discovery_engine_core/core/src/engine.rs

```diff
@@ -570,71 +575,75 @@ fn rank_stacks<'a>(
-                            let error = Error::Ranker(error);
-                            error!("{}", error);
-                            error
-                        })?;
-                document_from_article(article, id, embedding).map_err(|error| {
-                    let error = Error::Document(error);
+        let key_phrases = needy_stacks
+            .iter()
+            .any(|stack| stack.ops.needs_key_phrases())
+            .then(|| ranker.take_key_phrases(take_top))
+            .unwrap_or_default();
+
+        // Here we gather new documents for all relevant stacks, and put them into a vector.
+        // We don't update the stacks immediately, because we want to de-duplicate the documents
+        // across stacks first.
+        for stack in &mut needy_stacks {
+            let maybe_new_documents =
+                fetch_new_documents_for_stack(stack, ranker, &key_phrases, history).await;
+
+            match maybe_new_documents {
+                Err(Error::StackOpFailed(stack::Error::New(NewItemsError::NotReady))) => {
+                    ready_stacks -= 1;
+                    continue;
+                }
+                Err(error) => {
                     error!("{}", error);
-                    error
-                })
-            })
-            .partition_map::<Vec<_>, Vec<_>, _, _, _>(|result| match result {
-                Ok(document) => Either::Left(document),
-                Err(error) => Either::Right(error),
-            });
-        // only push an error if the articles aren't empty for other reasons and all articles failed
-        if articles_len > 0 && articles_errors.len() == articles_len {
-            errors.push(Error::Errors(articles_errors));
-            continue;
+                    errors.push(error);
+                    continue;
+                }
+                Ok(documents) => all_documents.extend(documents),
+            };
         }
+    }
+
+    // Since we need to de-duplicate not only the newly retrieved documents among themselves,
+    // but also consider the existing documents in all stacks (not just the needy ones), we extract
+    // them into `all_documents` here.
+    for stack in stacks.values_mut() {
+        all_documents.extend(stack.drain_documents());
+    }
+
+    // Separate stack documents (via the `stack` parameter) are not needed here, since they are
+    // already contained in `all_documents`.
+    all_documents = DuplicateFilter::apply(history, &[], all_documents);
+
+    // Finally, we can update the stacks with their respective documents. To do this, we
+    // have to group the fetched documents by `stack_id`, then `update` the stacks.
+    let documents_by_stack_id = all_documents.into_iter().group_by(|doc| doc.stack_id);
+    for (stack_id, documents_group) in &documents_by_stack_id {
+        let stack = stacks.get_mut(&stack_id).unwrap();
 
+        let documents: Vec<Document> = documents_group.collect();
```

Comment:
🎨 with `collect_vec()` from itertools we don't need to specify the type
```suggestion
        let documents = documents_group.collect_vec();
```